entity-service/src/main/kotlin/com/egm/stellio/entity/repository/Neo4jRepository.kt

```diff
@@ -255,48 +275,88 @@ class Neo4jRepository(
+            OPTIONAL MATCH (prop)-[:HAS_VALUE]->(propOfProp:Property)
+            WITH prop, propOfProp
+            OPTIONAL MATCH (prop)-[:HAS_OBJECT]->(relOfProp:Relationship)
+            DETACH DELETE prop, propOfProp, relOfProp
+            """.trimIndent()
+        else if (datasetId == null)
+            """
+            MATCH (entity:${subjectNodeInfo.label} { id: ${'$'}entityId })-[:HAS_VALUE]->(prop:Property { name: ${'$'}propertyName })
+            WHERE NOT EXISTS (prop.datasetId)
+            OPTIONAL MATCH (prop)-[:HAS_VALUE]->(propOfProp:Property)
+            WITH prop, propOfProp
+            OPTIONAL MATCH (prop)-[:HAS_OBJECT]->(relOfProp:Relationship)
+            DETACH DELETE prop, propOfProp, relOfProp
+            """.trimIndent()
+        else
+            """
+            MATCH (entity:${subjectNodeInfo.label} { id: ${'$'}entityId })-[:HAS_VALUE]->(prop:Property { name: ${'$'}propertyName, datasetId: ${'$'}datasetId})
             OPTIONAL MATCH (prop)-[:HAS_VALUE]->(propOfProp:Property)
             WITH prop, propOfProp
             OPTIONAL MATCH (prop)-[:HAS_OBJECT]->(relOfProp:Relationship)
             DETACH DELETE prop, propOfProp, relOfProp
             """.trimIndent()
 
         val parameters = mapOf(
-            "entityId" to entityId,
-            "propertyName" to propertyName
+            "entityId" to subjectNodeInfo.id,
+            "propertyName" to propertyName,
+            "datasetId" to datasetId?.toString()
         )
+
         val queryStatistics = session.query(query, parameters).queryStatistics()
         return queryStatistics.nodesDeleted
     }
 
-    fun deleteEntityRelationship(entityId: String, relationshipType: String): Int {
+    fun deleteEntityRelationship(subjectNodeInfo: SubjectNodeInfo, relationshipType: String, datasetId: URI? = null, deleteAll: Boolean = false): Int {
         /**
          * Delete :
          *
          * 1. the relationship
          * 2. the properties of the relationship
          * 3. the relationships of the relationship
          */
-        val query = """
-            MATCH (entity:Entity { id: ${'$'}entityId })-[:HAS_OBJECT]->(rel)-[:$relationshipType]->()
+        val query = if (deleteAll)
+            """
+            MATCH (entity:${subjectNodeInfo.label} { id: ${'$'}entityId })-[:HAS_OBJECT]->(rel:Relationship)-[:$relationshipType]->()
             OPTIONAL MATCH (rel)-[:HAS_VALUE]->(propOfRel:Property)
             WITH rel, propOfRel
             OPTIONAL MATCH (rel)-[:HAS_OBJECT]->(relOfRel:Relationship)
             DETACH DELETE rel, propOfRel, relOfRel
-        """.trimIndent()
+            """.trimIndent()
+        else if (datasetId == null)
+            """
+            MATCH (entity:${subjectNodeInfo.label} { id: ${'$'}entityId })-[:HAS_OBJECT]->(rel:Relationship)-[:$relationshipType]->()
+            WHERE NOT EXISTS (rel.datasetId)

```

Comment:
deleteAll is this query without this specific line, couldn't we use that to reduce redundancy ?